oak-core/src/main/java/org/apache/jackrabbit/oak/security/user/ImpersonationImpl.java

```diff
@@ -173,13 +182,48 @@ private void updateImpersonatorNames(@NotNull Tree userTree, @NotNull Set<String
     private boolean isAdmin(@NotNull Principal principal) {
```

Comment:
I would also change the name if this method to better reflect what is currently doing, as it isn't just checking if the principal is an admin anymore, but wether it can impersonate freely.

Reply:
Thanks for suggestion, the isAdmin now performs only the admin check. 
+ The unnecessary impersonation related checks might've also impacted the performance.